### PR TITLE
Hoang/check num of val to rebalance

### PR DESCRIPTION
**Pull request by hoangdv2429**
Closes: #419

## Context and purpose of the change

I also added test for code written in validateBasic of stakeibc, however in order to get it to pass. You will have to change bech32prefix from cosmos to stride (I wrote some comments describe the detail in the code)

## Brief Changelog

Move num val checking from msg_server to validateBasic and add test for it.

## Author's Checklist

I have...

- [x] Run and PASSED locally all Stride integration tests
- [x] If the change is contentful, I either:
    - [x] Added a new unit test OR 
    - [x] Added test cases to existing unit tests
- [ ] OR this change is a trivial rework / code cleanup without any test coverage

If skipped any of the tests above, explain.

I also added test for code written in validateBasic of stakeibc, however in order to get it to pass. You will have to change bech32prefix from cosmos to stride (I wrote some comments describe the detail in the code)

## Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] manually tested (if applicable)
- [ ] confirmed the author wrote unit tests for new logic
- [ ] reviewed documentation exists and is accurate

## Documentation and Release Note

  - [ ] Does this pull request introduce a new feature or user-facing behavior changes? 
  - [ ] Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`?
  - [ ] This pull request updates existing proto field values (and require a backend and frontend migration)? 
  - [ ] Does this pull request change existing proto field names (and require a frontend migration)?
  How is the feature or change documented? 
      - [ ] not applicable
      - [ ] jira ticket `XXX` 
      - [ ] specification (`x/<module>/spec/`) 
      - [ ] README.md 
      - [ ] not documented

### CODE REVIEW

**Comment by hieuvubk** [comment]
@hoangdv2429 hey could u run `make test-unit` and check, seem some test are failing

**Comment by hoangdv2429** [comment]
> @hoangdv2429 hey could u run `make test-unit` and check, seem some test are failing

I updated it with app params. all test should pass now.

File: x/stakeibc/types/message_rebalance_validators_test.go [hunk 1]
```diff
@@ -27,6 +32,27 @@ func TestMsgRebalanceValidators_ValidateBasic(t *testing.T) {
 				Creator: sample.AccAddress(),
 			},
 			err: sdkerrors.ErrInvalidAddress,
+		}, {
```
Comment (riley-stride) [inline x/stakeibc/types/message_rebalance_validators_test.go:35]: @sampocs do you think we should move these tests back to `x/stakeibc/keeper/msg_server_rebalance_validators_test.go` so that we have consistency in our testing approach across the codebase?

File: x/stakeibc/types/message_rebalance_validators.go [hunk 2]
```diff
@@ -10,6 +10,8 @@ import (
 )
 
 const TypeMsgRebalanceValidators = "rebalance_validators"
+const MIN_NUM_REBALANCE = 1
+const MAX_NUM_REBALANCE = 10
```
Comment (riley-stride) [inline x/stakeibc/types/message_rebalance_validators.go:14]: If you set these constants here, shouldn't you reference them on line 55, rather than leaving the hardcoded values?

**Review by riley-stride — CHANGES_REQUESTED** [review]
Thanks for opening this PR. A few questions and requests. 

1) Defining MIN and MAX as constants makes sense, though I think making them params would give the protocol a bit more flexibility. 

2) Would like to get @sampocs's review on this and his opinion on the testing framework choice.

3) Seems like you are still using the hardcoded values in the ValidateBasic check in `message_rebalance_validators.go`. 

4) Why are you adding the `app/params` directory?

**Comment by sampocs** [comment]
@hoangdv2429 @riley-stride 
Considering the check is now in Validate basic, I think moving it to the testing in types makes sense since our framework is meant for keeper functions. That said, I'm not sure all those files are necessary to just to get the address prefix to work. 

I handled something similar to this in a separate branch that hasn't been merged yet. I'll pull it out into a new PR and you can pull that into this.

**Comment by sampocs** [comment]
For what it's worth, my philosophy on testing (since it's come up across a couple of PRs): 

The "array-of-test-cases" approach is good for simple go tests like those in types and ValidateBasic, _where the input to each test case is small._

However, when it comes to our keeper functions, there is often a lot of state that needs to be mocked out and each test can have various other conditions and branches (for instance, you can imagine an ICA unit test where you might want to close a channel in the middle of a test). In those scenarios, I think the "array-of-test-cases" approach quickly becomes very messy as it's difficult to generalize logic across each test case tests. Plus, the test function itself becomes bloated, which makes it hard to follow what's going on. So in those scenarios, I think the "function-per-test-case" approach makes sense as you see with most of our keeper functions.

All that said, I support the decision to move these tests over in this PR! Curious to hear your guys' thoughts as well!

**Comment by hieuvubk** [comment]
@riley-stride @sampocs Thank u for leaving comment!
Yes I totally agree that we should move MIN & MAX to params.
@sampocs Thank u for helper func its help me so much

**Comment by hieuvubk** [comment]
Also if we use MIN & MAX as params. We have to move the check from `ValidateBasic` to keeper func (of course we can't get module params in `types`)
Pls lmk what u think @riley-stride @sampocs

**Comment by hoangdv2429** [comment]
@sampocs @riley-stride  can you review this PR ? If there is anything wrong with it we'll try to change it asap. Thank you!

**Comment by sampocs** [comment]
Sorry for the delay on this! Would you mind merging in main and resolving the conflicts and then I'll give it another review!

**Comment by hoangdv2429** [comment]
@sampocs sorry I didn't see your message earlier, I merged and resolved all conflicts. Can you have a look

File: proto/stride/stakeibc/params.proto [hunk 3]
```diff
@@ -33,5 +33,7 @@ message Params {
   uint64 safety_max_redemption_rate_threshold = 15;
   uint64 ibc_transfer_timeout_nanos = 16;
   uint64 safety_num_validators = 17;
-  uint64 safety_max_slash_percent = 18;
+  uint64 min_num_rebalance = 18;
+  uint64 max_num_rebalance = 19;
+  uint64 safety_max_slash_percent = 20;
```
Comment (sampocs) [inline proto/stride/stakeibc/params.proto:38]: Lets keep the slash percent as the same index to make the migration cleaner
```suggestion
  uint64 safety_max_slash_percent = 18;
  uint64 min_num_rebalance = 19;
  uint64 max_num_rebalance = 20;
```

File: x/stakeibc/keeper/msg_server_rebalance_validators.go [hunk 4]
```diff
@@ -26,15 +26,14 @@ func (k msgServer) RebalanceValidators(goCtx context.Context, msg *types.MsgReba
...
-		k.Logger(ctx).Error(fmt.Sprintf("Invalid number of validators to rebalance %d", maxNumRebalance))
-		return nil, types.ErrInvalidNumValidator
-	}
-	if maxNumRebalance > 4 {
-		k.Logger(ctx).Error(fmt.Sprintf("Invalid number of validators to rebalance %d", maxNumRebalance))
-		return nil, types.ErrInvalidNumValidator
+
+	params := k.GetParams(ctx)
+	if (msg.NumRebalance < params.MinNumRebalance) || (msg.NumRebalance > params.MaxNumRebalance) {
+		return nil, fmt.Errorf("invalid number of validators to rebalance (%d)", msg.NumRebalance)
```
Comment (sampocs) [inline x/stakeibc/keeper/msg_server_rebalance_validators.go:32]:
```suggestion
		return nil, errorsmod.Wrapf(types.ErrInvalidNumValidator, "invalid number of validators to rebalance (%d)", msg.NumRebalance)
```

File: x/stakeibc/keeper/msg_server_rebalance_validators.go [hunk 5]
```diff
@@ -26,15 +26,14 @@ func (k msgServer) RebalanceValidators(goCtx context.Context, msg *types.MsgReba
...
 	}
+	
+	maxNumRebalance := cast.ToInt(msg.NumRebalance)
```
Comment (sampocs) [inline x/stakeibc/keeper/msg_server_rebalance_validators.go:35]: You can remove this line and change the forloop below (where maxNumRebalance is used) to 

```
for i := uint64(1); i <= maxNumRebalance; i++ {
```

File: x/stakeibc/types/params.go [hunk 6]
```diff
@@ -27,6 +27,8 @@ var (
 	DefaultMaxStakeICACallsPerEpoch         uint64 = 100
 	DefaultIBCTransferTimeoutNanos          uint64 = 1800000000000 // 30 minutes
 	DefaultSafetyNumValidators              uint64 = 35
+	DefaultMinNumRebalance					uint64 = 1
```
Comment (sampocs) [inline x/stakeibc/types/params.go:30]: can you double check the linting in this file? the spacing seems off

File: x/stakeibc/types/message_rebalance_validators.go [hunk 7]
```diff
@@ -52,7 +50,7 @@ func (msg *MsgRebalanceValidators) ValidateBasic() error {
 		return err
 	}
 	if (msg.NumRebalance < 1) || (msg.NumRebalance > 10) {
```
Comment (sampocs) [inline x/stakeibc/types/message_rebalance_validators.go:52]: This should use the same parameters that were just added

**Review by sampocs — COMMENTED** [review]
Good stuff! Some nits comments below. One large point:

The main issue for this ticket was that checks in validate basic and the msg_server_rebalance_vals should be consistent, but if the min/max are stored as params in the store, there's no way to grab them from validate basic. 

I think the two options are:
1. Remove the check from validate basic 
2. Use a true golang constant instead of a module param (as recommended in the audit)

I think my vote is to go with option 2 as it's little simpler and then we also don't need a migration for this param. Curious to here @riley-stride's thoughts

File: x/stakeibc/types/params.go (hunk 6 above)
Comment (hoangdv2429) [reply x/stakeibc/types/params.go:30]: idk why but it's look fine on my side though 
image

**Comment by hoangdv2429** [comment]
@sampocs I made the changes as suggested. Can you have a look ?

**Comment by sampocs** [comment]
Closing in favor of #630 so the min/max variables are represented as constants instead of params.
The audit report suggested just checking in one place (validate basic) and validate basic does not have access to the store, so it makes more sense to just use constants.
This also prevents the need for a store migration
